From 529dfb8cf9c77fb21d7d7acf77bb69894c5e5eb5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Cl=C3=A9ment=20P=C3=A9ron?= <peron.clem@gmail.com>
Date: Tue, 27 Dec 2022 21:00:30 +0100
Subject: 0002-ubnt-min-lease-time-6-sec.patch


diff --git a/networking/udhcp/dhcpc.c b/networking/udhcp/dhcpc.c
index d5ebcb4b4..f3e4bd737 100644
--- a/networking/udhcp/dhcpc.c
+++ b/networking/udhcp/dhcpc.c
@@ -1821,8 +1821,8 @@ int udhcpc_main(int argc UNUSED_PARAM, char **argv)
 				 */
 				if (lease_remaining < 0) /* signed overflow? */
 					lease_remaining = INT_MAX;
-				if (lease_remaining < 30)
-					lease_remaining = 30;
+				if (lease_remaining < 6)
+					lease_remaining = 6;
 				requested_ip = packet.yiaddr;
 #if ENABLE_FEATURE_UDHCPC_ARPING
 				if (opt & OPT_a) {
